path_finding.asciidoc

```diff
@@ -199,19 +210,65 @@ But wait: Selena does know *some* channel balances! She knows the balances of th
+
 [[channel_graph_3]]
 .Channel graph fees and other channel metrics
 image::images/channel_graph_3.png[]
 
+The fee and timelock information are very important not just as path selection metrics. As we saw in <<onion_routing>>, the sender needs to add up fees and timelocks (cltv_expiry_delta) at each hop to make the onion. The process of calculating fees happens from the recipient to the sender *backwards* along the path, because each intermediary hop expects an incoming HTLC with higher amount and expiry timelock than the outgoing HTLC they will send to the next hop. So, for example, if Bob wants 1000 satoshis in fees and 30 blocks of expiry timelock delta, to send a payment to Rashid, then that amount and expiry delta must be added to the HTLC _from Alice_.
+
+It is also important to note that a channel must have liquidity that is sufficient not only for the payment amount but also for the cumulative fees of all the subsequent hops. That means that if we are paying 100,000 satoshis to Rashid and a total of 5000 satoshis in fees for the intermediary hops, the liquidity of the first channel must exceed 105,000 satoshis. We need to know fees because only paths that have sufficient liquidity for *both payment and all fees* will be considered.
 
 ==== Finding candidate paths
 
-==== Selecting a path, ordering of paths
+Finding a suitable path through a directed graph like this is a well-studied computer science problem (known broadly as the "Shortest Path problem"), which can be solved by a variety of algorithms depending on the desired optimization and resource constraints.
+
+The most famous algorithm solving this problem was invented by Dutch mathematician E. W. Dijkstra in 1956, known simply as https://en.wikipedia.org/wiki/Dijkstra's_algorithm[_Dijkstra's Algorithm_]. In addition to the original Dijkstra algorithm, there are many variations and optimizations, such as https://en.wikipedia.org/wiki/A*_search_algorithm[A* ("A-star")], which is a heuristic-based algorithm.
+
+As mentioned previously, the "search" must be applied _backwards_ to account for fees that are accumulated from recipient to sender. Thus, Dijkstra, A* or some other algorithm would search for a path from the recipient to the sender, using fees, estimated liquidity, timelock delta (or some combination) as a cost function for each hop.
 
 ==== Payment delivery (Trial-and-error loop)
 
-==== Updating the channel graph
+Once Selena has a selected path, her node starts the trial-and-error loop, by constructing the HTLCs, building the onion and attempting delivery of the payment. For each attempt, there are three possible outcomes:
+
+- A successful result (+update_fulfill_htlc+)
+- An error (+update_fail_htlc+)
+- A "stuck" payment with no response (neither success, nor failure)
```

Comment:
here we should differentiate as there are two forms of stuck payments (which we for now cannot differentiate on a protocol level) when we do MPP a path can be arrived at the recipient and all the HTLCs are locked in but the path will not settle yet as other onions need to arrive to achieve the full amount. That is what I would call an `expected` stuck payment or a locked htlc. then of course ther are the stuck payments (which I guess you had in mind when you wrote this) that are due to poor / hostile behaviour of routing nodes that just delay forwarding the htlcs or errors.